openstack/compute/v2/servers/requests.go

```diff
@@ -277,6 +281,12 @@ func (opts CreateOpts) ToServerCreateMap() (map[string]interface{}, error) {
 
 // Create requests a server to be provisioned to the user in the current tenant.
 func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r CreateResult) {
+	err := gophercloud.ValidateMicroversionRequest(client.Microversion, opts)
```

Comment:
Yep, totally possible. I debated doing this or bundling it with `BuildRequestBody`. Decided to go this route to see how it looked. The benefit is that it's apparent this is a microversion-enabled request. I think there's merit there.

Reply:
I've thought about this more and still prefer having the check here. In addition to being more visible that it's a microversion-enabled request, having it closer to the core (either `Request` or `BuildRequestBody`) will have it run for _all_ requests - even for those services that don't use microversions.

Reply:
> will have it run for all requests - even for those services that don't use microversions.

That's a good point. I'm not sure I want it at all though. Either an error is returned here for an invalid microversion or it's returned by OpenStack for an invalid microversion. I'm not sure there's not much a user can act on programatically there.

Reply:
> or it's returned by OpenStack for an invalid microversion

This isn't guaranteed. I've seen cases of both an error being returned and the field being silently ignored. I'd prefer to inject a guaranteed error rather than have something silently ignored.

> I'm not sure there's not much a user can act on programatically there.

Isn't this similar to the `required`, `xor`, etc tags?

Assuming the application is exposing the ability for a user to set their client's microversion, the user can alter their version or input.

Reply:
> Isn't this similar to the required, xor, etc tags?

Yep, those are a convenience, too. But they're also relatively scarce. What I don't like about this approach is that these calls will be littered throughout the library. I'll keep thinking about it.

Reply:
That's fair. In that regard, I'd rather have this integrated with `BuildRequestBody`, `BuildQueryString`, etc. IMO, giving the user a confident block that their request will not be what they expect it is the important thing. Burning some extra cycles looking for non-existent `mv` tags is minor in comparison.

Reply:
I'd leave it as-is for now. I'd like to think about it a bit more

Reply:
After thinking about it some more, I'm convinced this is the correct way to do it. Maybe it should go in the `Build*Opts` function and only called if the user provides a microversioned field.

Reply:
To clarify: by `Build*Opts` do you mean the core functions in `params.go`?

I realized that in order to do that, we need to pass the value of `client.Microversion` somehow: either adding an input parameter to `ToServerCreateMap()` or adding an unexported field to `CreateOpts`. Any suggestion on the best way to go?

And then the `BuildRequestBody` and `BuildQueryString` will also need a new input parameter.

I'm OK doing all of that, but I now realize why I sided on this method to prototype -- less disruptive changes :)

Reply:
Ugh, I mis-typed that. I meant putting the call in, say, the `ToServerCreateMap`, creating a `map[string]string` that maps fields to microversions (or a slice of a new top-level struct with fields like `Name`, `Min` and `Max`), and then passes only the microversion fields to be checked and only if the user provides them.

Reply:
:thinking: 

Interesting. I'll give that a shot :)

Reply:
The checking-fields bit may not be unmaintainable. If all the fields are going to end up having microversions, it's probably more practical to pass in the whole `*Opts` like you're doing now.

Reply:
Yeah, given the new fields will have to go in `*Opts` anyway, and if we imagine Gophercloud 2-3 years from now with a good amount of new fields, then using an `mv` tag is the most succinct way to go.

Creating a top-level `Microversion` struct which can be used to describe the amended fields in a per-package `microversions.go` file provides a dedicated schema of sorts, which could be interesting... but it's either duplicating data or dividing the required data into two sources.